File: sdk/typescript/src/resources/mail.ts
// Mail capability — Mashgate core primitive (ADR-0019).
//
// Mirrors `mail.v1.MailService` from
// mashgate/contracts/proto/v1/mail.proto, exposed over the gateway as REST
// via google.api.http transcoding.
//
// Auth: pass an end-user JWT for self-service mailbox operations (read/write
// scope), or an admin/service-account token for tenant-wide operations
// (mail:admin scope).
//
// Events (subscribe via webhooks): mail.received / mail.sent /
// mail.delivered / mail.bounced — see contracts/events/mail.*.json.

import type { MashgateClient } from "../client.js";

// ── Enum-like string types from mail.proto ──────────────────────────────────

export const MessageFolder = {
  Inbox: "MESSAGE_FOLDER_INBOX",
  Sent: "MESSAGE_FOLDER_SENT",
  Drafts: "MESSAGE_FOLDER_DRAFTS",
  Spam: "MESSAGE_FOLDER_SPAM",
  Trash: "MESSAGE_FOLDER_TRASH",
} as const;
export type MessageFolder = (typeof MessageFolder)[keyof typeof MessageFolder];

export const MailboxStatus = {
  Active: "MAILBOX_STATUS_ACTIVE",
  Frozen: "MAILBOX_STATUS_FROZEN",
  Closed: "MAILBOX_STATUS_CLOSED",
} as const;
export type MailboxStatus = (typeof MailboxStatus)[keyof typeof MailboxStatus];

export const DomainStatus = {
  Pending: "DOMAIN_STATUS_PENDING",
  Active: "DOMAIN_STATUS_ACTIVE",
  Suspended: "DOMAIN_STATUS_SUSPENDED",
} as const;
export type DomainStatus = (typeof DomainStatus)[keyof typeof DomainStatus];

export const SendStatus = {
  Queued: "SEND_STATUS_QUEUED",
  Sent: "SEND_STATUS_SENT",
  Delivered: "SEND_STATUS_DELIVERED",
  Failed: "SEND_STATUS_FAILED",
} as const;
export type SendStatus = (typeof SendStatus)[keyof typeof SendStatus];

// ── Domain types ────────────────────────────────────────────────────────────

export interface Mailbox {
  mailbox_id: string;
  tenant_id: string;
  subject_id: string;
  email: string;
  display_name?: string;
  status: MailboxStatus;
  quota_bytes: number;
  used_bytes: number;
  created_at: string;
  updated_at: string;
}

export interface MessagePreview {
  message_id: string;
  tenant_id: string;
  mailbox_id: string;
  from: string;
  to: string[];
  subject: string;
  preview?: string;
  received_at: string;
  read: boolean;
  folder: MessageFolder;
  labels?: string[];
  has_attachments?: boolean;
}

export interface MessageAttachment {
  attachment_id: string;
  filename: string;
  content_type: string;
  size_bytes: number;
  /** Signed URL (Mashgate storage-service). Не embedded в payload. */
  url: string;
}

export interface Message {
  preview: MessagePreview;
  body_text?: string;
  body_html?: string;
  headers?: Record<string, string>;
  attachments?: MessageAttachment[];
  cc?: string[];
  bcc?: string[];
  sent_at?: string;
  in_reply_to?: string;
}

export interface Domain {
  domain_id: string;
  tenant_id: string;
  name: string;
  status: DomainStatus;
  dkim_selector?: string;
  /** Public key для копи-паста в DNS TXT-record. */
  dkim_public_key?: string;
  mx_records: string[];
  spf_record?: string;
  dmarc_record?: string;
  /** Что не настроено в DNS — список записей которые VerifyDomain не нашёл. */
  verification_errors?: string[];
  created_at: string;
  last_verified_at?: string;
}

// ── Request shapes ──────────────────────────────────────────────────────────

export interface ListMessagesQuery {
  folder?: MessageFolder;
  limit?: number;
  cursor?: string;
  /** Admin only (mail:admin scope): чужой mailbox. Без этого — собственный. */
  mailbox_id_override?: string;
}

export interface ListMessagesResponse {
  items: MessagePreview[];
  next_cursor?: string;
  total?: number;
}

export interface SendMessageRequest {
  to: string[];
  cc?: string[];
  bcc?: string[];
  subject: string;
  body_text?: string;
  body_html?: string;
  headers?: Record<string, string>;
  /** ID файлов pre-uploaded через Mashgate storage-service. */
  attachment_ids?: string[];
  /** Message-ID письма-родителя для threading. */
  in_reply_to?: string;
  idempotency_key?: string;
}

export interface SendMessageResponse {
  message_id: string;
  status: SendStatus;
  note?: string;
  queued_at: string;
}

export interface UpdateMessageRequest {
  read?: boolean;
  folder?: MessageFolder;
  labels?: string[];
}

export interface ListMailboxesQuery {
  status?: MailboxStatus;
  limit?: number;
  cursor?: string;
}

export interface ListMailboxesResponse {
  items: Mailbox[];
  next_cursor?: string;
}

export interface CreateMailboxRequest {
  subject_id: string;
  email: string;
  display_name?: string;
  /** 0 = дефолтная квота тенанта. */
  quota_bytes?: number;
  idempotency_key?: string;
}

export interface ListDomainsQuery {
  status?: DomainStatus;
}

export interface ListDomainsResponse {
  items: Domain[];
}

export interface CreateDomainRequest {
  /** FQDN, e.g. "mail.entry-i.com". */
  name: string;
}

export interface RotateDKIMRequest {
  /** 1024 / 2048. По умолчанию 2048. */
  key_bits?: number;
}

// ── Resource ────────────────────────────────────────────────────────────────

export class MailResource {
  constructor(private readonly client: MashgateClient) {}

  // ── User-facing (mail:read / mail:write) ─────────────────────────────

  /** Mailbox текущего пользователя (по subject из JWT). */
  getMyMailbox(): Promise<Mailbox> {
    return this.client.request<Mailbox>("GET", "/v1/mail/mailboxes/me");
  }

  /** Список писем в папке (cursor pagination). */
  listMessages(query?: ListMessagesQuery): Promise<ListMessagesResponse> {
    return this.client.request<ListMessagesResponse>("GET", "/v1/mail/messages", {
      query: {
        folder: query?.folder,
        limit: query?.limit,
        cursor: query?.cursor,
        mailbox_id_override: query?.mailbox_id_override,
      },
    });
  }

  /** Полное письмо (body + headers + attachment links). */
  getMessage(messageId: string): Promise<Message> {
    return this.client.request<Message>("GET", `/v1/mail/messages/${messageId}`);
  }

  /**
   * Отправить письмо. Идемпотентно через `idempotency_key` — повторный
   * вызов с тем же ключом возвращает тот же `message_id`. Реальная SMTP-доставка
   * происходит асинхронно; статус потом приходит через события
   * `mail.sent` / `mail.delivered` / `mail.bounced`.
   */
  sendMessage(req: SendMessageRequest): Promise<SendMessageResponse> {
    return this.client.request<SendMessageResponse>("POST", "/v1/mail/messages", {
      body: req,
    });
  }

  /** Обновить флаги (read/unread, folder, labels). Только переданные поля меняются. */
  updateMessage(messageId: string, req: UpdateMessageRequest): Promise<Message> {
    return this.client.request<Message>("PATCH", `/v1/mail/messages/${messageId}`, {
      body: req,
    });
  }

  /**
   * Soft-delete — moves to TRASH. Если письмо уже в TRASH и `hardDelete=true` —
   * удаляется навсегда.
   */
  deleteMessage(messageId: string, hardDelete = false): Promise<void> {
    return this.client.request<void>("DELETE", `/v1/mail/messages/${messageId}`, {
      query: { hard_delete: hardDelete },
    });
  }

  // ── Admin-facing (mail:admin) ────────────────────────────────────────

  /** Все ящики тенанта (admin). */
  listMailboxes(query?: ListMailboxesQuery): Promise<ListMailboxesResponse> {
    return this.client.request<ListMailboxesResponse>("GET", "/v1/mail/mailboxes", {
      query: {
        status: query?.status,
        limit: query?.limit,
        cursor: query?.cursor,
      },
    });
  }

  /** Создать ящик (admin). Email должен быть в одном из активных доменов тенанта. */
  createMailbox(req: CreateMailboxRequest): Promise<Mailbox> {
    return this.client.request<Mailbox>("POST", "/v1/mail/mailboxes", { body: req });
  }

  /** Домены тенанта. */
  listDomains(query?: ListDomainsQuery): Promise<ListDomainsResponse> {
    return this.client.request<ListDomainsResponse>("GET", "/v1/mail/domains", {
      query: { status: query?.status },
    });
  }

  /**
   * Зарегистрировать домен (admin). Возвращает Domain в статусе `pending` с
   * сгенерированным `dkim_public_key` — пользователь должен прописать TXT-записи
   * в свой DNS (DKIM/SPF/DMARC), затем вызвать `verifyDomain`.
   */
  createDomain(req: CreateDomainRequest): Promise<Domain> {
    return this.client.request<Domain>("POST", "/v1/mail/domains", { body: req });
  }

  /**
   * Проверить DNS-записи домена. Если все на месте — статус становится
   * `active` и домен может отправлять/принимать почту. Иначе возвращает
   * `verification_errors` со списком проблем.
   */
  verifyDomain(domainId: string): Promise<Domain> {
    return this.client.request<Domain>(
      "POST",
      `/v1/mail/domains/${domainId}/verify`,
      { body: {} },
    );
  }

  /**
   * Сгенерировать новый DKIM-ключ. Старый селектор остаётся активным до
   * следующей ротации (через ~30 дней) — это grace period для успешной
   * пропагации DNS у получателей.
   */
  rotateDKIM(domainId: string, req?: RotateDKIMRequest): Promise<Domain> {
    return this.client.request<Domain>(
      "POST",
      `/v1/mail/domains/${domainId}/dkim/rotate`,
      { body: { key_bits: req?.key_bits ?? 2048 } },
    );
  }
}

File: sdk/go/mail.go
package mashgate

import (
	"context"
	"fmt"
	"net/url"
	"strconv"
	"time"
)

// Mail capability — Mashgate core primitive (ADR-0019). Mirrors mail.v1.MailService
// from mashgate/contracts/proto/v1/mail.proto, exposed over the gateway as REST
// via google.api.http transcoding.
//
// Auth: pass a user JWT for self-service mailbox operations (mail:read /
// mail:write scope), or admin/service-account credentials for tenant-wide
// operations (mail:admin scope).
//
// Events (subscribe via webhooks): mail.received / mail.sent / mail.delivered /
// mail.bounced — see contracts/events/mail.*.json.

// ────────────────────────────────────────────────────────────────────────────
// Enum-like string types from mail.proto
// ────────────────────────────────────────────────────────────────────────────

type MessageFolder string

const (
	MessageFolderInbox  MessageFolder = "MESSAGE_FOLDER_INBOX"
	MessageFolderSent   MessageFolder = "MESSAGE_FOLDER_SENT"
	MessageFolderDrafts MessageFolder = "MESSAGE_FOLDER_DRAFTS"
	MessageFolderSpam   MessageFolder = "MESSAGE_FOLDER_SPAM"
	MessageFolderTrash  MessageFolder = "MESSAGE_FOLDER_TRASH"
)

type MailboxStatus string

const (
	MailboxStatusActive MailboxStatus = "MAILBOX_STATUS_ACTIVE"
	MailboxStatusFrozen MailboxStatus = "MAILBOX_STATUS_FROZEN"
	MailboxStatusClosed MailboxStatus = "MAILBOX_STATUS_CLOSED"
)

type DomainStatus string

const (
	DomainStatusPending   DomainStatus = "DOMAIN_STATUS_PENDING"
	DomainStatusActive    DomainStatus = "DOMAIN_STATUS_ACTIVE"
	DomainStatusSuspended DomainStatus = "DOMAIN_STATUS_SUSPENDED"
)

type SendStatus string

const (
	SendStatusQueued    SendStatus = "SEND_STATUS_QUEUED"
	SendStatusSent      SendStatus = "SEND_STATUS_SENT"
	SendStatusDelivered SendStatus = "SEND_STATUS_DELIVERED"
	SendStatusFailed    SendStatus = "SEND_STATUS_FAILED"
)

// ────────────────────────────────────────────────────────────────────────────
// Domain types
// ────────────────────────────────────────────────────────────────────────────

type Mailbox struct {
	MailboxID   string        `json:"mailbox_id"`
	TenantID    string        `json:"tenant_id"`
	SubjectID   string        `json:"subject_id"`
	Email       string        `json:"email"`
	DisplayName string        `json:"display_name,omitempty"`
	Status      MailboxStatus `json:"status"`
	QuotaBytes  int64         `json:"quota_bytes"`
	UsedBytes   int64         `json:"used_bytes"`
	CreatedAt   time.Time     `json:"created_at"`
	UpdatedAt   time.Time     `json:"updated_at"`
}

type MailMessagePreview struct {
	MessageID      string        `json:"message_id"`
	TenantID       string        `json:"tenant_id"`
	MailboxID      string        `json:"mailbox_id"`
	From           string        `json:"from"`
	To             []string      `json:"to"`
	Subject        string        `json:"subject"`
	Preview        string        `json:"preview,omitempty"`
	ReceivedAt     time.Time     `json:"received_at"`
	Read           bool          `json:"read"`
	Folder         MessageFolder `json:"folder"`
	Labels         []string      `json:"labels,omitempty"`
	HasAttachments bool          `json:"has_attachments,omitempty"`
}

type MailAttachment struct {
	AttachmentID string `json:"attachment_id"`
	Filename     string `json:"filename"`
	ContentType  string `json:"content_type"`
	SizeBytes    int64  `json:"size_bytes"`
	URL          string `json:"url"` // signed URL via Mashgate storage-service
}

type MailMessage struct {
	Preview     MailMessagePreview `json:"preview"`
	BodyText    string             `json:"body_text,omitempty"`
	BodyHTML    string             `json:"body_html,omitempty"`
	Headers     map[string]string  `json:"headers,omitempty"`
	Attachments []MailAttachment   `json:"attachments,omitempty"`
	CC          []string           `json:"cc,omitempty"`
	BCC         []string           `json:"bcc,omitempty"`
	SentAt      *time.Time         `json:"sent_at,omitempty"`
	InReplyTo   string             `json:"in_reply_to,omitempty"`
}

type MailDomain struct {
	DomainID            string       `json:"domain_id"`
	TenantID            string       `json:"tenant_id"`
	Name                string       `json:"name"`
	Status              DomainStatus `json:"status"`
	DKIMSelector        string       `json:"dkim_selector,omitempty"`
	DKIMPublicKey       string       `json:"dkim_public_key,omitempty"`
	MXRecords           []string     `json:"mx_records"`
	SPFRecord           string       `json:"spf_record,omitempty"`
	DMARCRecord         string       `json:"dmarc_record,omitempty"`
	VerificationErrors  []string     `json:"verification_errors,omitempty"`
	CreatedAt           time.Time    `json:"created_at"`
	LastVerifiedAt      *time.Time   `json:"last_verified_at,omitempty"`
}

// ────────────────────────────────────────────────────────────────────────────
// Request / Response types
// ────────────────────────────────────────────────────────────────────────────

type ListMailMessagesQuery struct {
	Folder            MessageFolder
	Limit             int
	Cursor            string
	MailboxIDOverride string // mail:admin only
}

type ListMailMessagesResponse struct {
	Items      []MailMessagePreview `json:"items"`
	NextCursor string               `json:"next_cursor,omitempty"`
	Total      int                  `json:"total,omitempty"`
}

type SendMailRequest struct {
	To             []string          `json:"to"`
	CC             []string          `json:"cc,omitempty"`
	BCC            []string          `json:"bcc,omitempty"`
	Subject        string            `json:"subject"`
	BodyText       string            `json:"body_text,omitempty"`
	BodyHTML       string            `json:"body_html,omitempty"`
	Headers        map[string]string `json:"headers,omitempty"`
	AttachmentIDs  []string          `json:"attachment_ids,omitempty"`
	InReplyTo      string            `json:"in_reply_to,omitempty"`
	IdempotencyKey string            `json:"idempotency_key,omitempty"`
}

type SendMailResponse struct {
	MessageID string     `json:"message_id"`
	Status    SendStatus `json:"status"`
	Note      string     `json:"note,omitempty"`
	QueuedAt  time.Time  `json:"queued_at"`
}

type UpdateMailMessageRequest struct {
	Read   *bool         `json:"read,omitempty"`
	Folder MessageFolder `json:"folder,omitempty"`
	Labels []string      `json:"labels,omitempty"`
}

type ListMailboxesQuery struct {
	Status MailboxStatus
	Limit  int
	Cursor string
}

type ListMailboxesResponse struct {
	Items      []Mailbox `json:"items"`
	NextCursor string    `json:"next_cursor,omitempty"`
}

type CreateMailboxRequest struct {
	SubjectID      string `json:"subject_id"`
	Email          string `json:"email"`
	DisplayName    string `json:"display_name,omitempty"`
	QuotaBytes     int64  `json:"quota_bytes,omitempty"`
	IdempotencyKey string `json:"idempotency_key,omitempty"`
}

type ListMailDomainsResponse struct {
	Items []MailDomain `json:"items"`
}

type CreateMailDomainRequest struct {
	Name string `json:"name"`
}

type RotateDKIMRequest struct {
	KeyBits int `json:"key_bits,omitempty"`
}

// ────────────────────────────────────────────────────────────────────────────
// MailClient
// ────────────────────────────────────────────────────────────────────────────

type MailClient struct {
	c *Client
}

// ── User-facing (mail:read / mail:write) ────────────────────────────────────

// GetMyMailbox returns the mailbox for the authenticated subject.
func (m *MailClient) GetMyMailbox(ctx context.Context) (*Mailbox, error) {
	var out Mailbox
	if err := m.c.do(ctx, "GET", "/v1/mail/mailboxes/me", nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// ListMessages returns a page of message previews from the authenticated
// subject's mailbox.
func (m *MailClient) ListMessages(ctx context.Context, q ListMailMessagesQuery) (*ListMailMessagesResponse, error) {
	v := url.Values{}
	if q.Folder != "" {
		v.Set("folder", string(q.Folder))
	}
	if q.Limit > 0 {
		v.Set("limit", strconv.Itoa(q.Limit))
	}
	if q.Cursor != "" {
		v.Set("cursor", q.Cursor)
	}
	if q.MailboxIDOverride != "" {
		v.Set("mailbox_id_override", q.MailboxIDOverride)
	}
	path := "/v1/mail/messages"
	if encoded := v.Encode(); encoded != "" {
		path += "?" + encoded
	}
	var out ListMailMessagesResponse
	if err := m.c.do(ctx, "GET", path, nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// GetMessage returns a single message with body and headers.
func (m *MailClient) GetMessage(ctx context.Context, messageID string) (*MailMessage, error) {
	var out MailMessage
	path := fmt.Sprintf("/v1/mail/messages/%s", url.PathEscape(messageID))
	if err := m.c.do(ctx, "GET", path, nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// SendMessage queues a message for delivery. Idempotent via IdempotencyKey:
// the same key on the same tenant returns the original message_id.
func (m *MailClient) SendMessage(ctx context.Context, req SendMailRequest) (*SendMailResponse, error) {
	var out SendMailResponse
	if err := m.c.do(ctx, "POST", "/v1/mail/messages", req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// UpdateMessage patches mutable flags (read/folder/labels) on a message.
// Pass only the fields you want to change.
func (m *MailClient) UpdateMessage(ctx context.Context, messageID string, req UpdateMailMessageRequest) (*MailMessage, error) {
	var out MailMessage
	path := fmt.Sprintf("/v1/mail/messages/%s", url.PathEscape(messageID))
	if err := m.c.do(ctx, "PATCH", path, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// DeleteMessage moves a message to TRASH. If hardDelete=true and the message
// is already in TRASH, it is permanently removed.
func (m *MailClient) DeleteMessage(ctx context.Context, messageID string, hardDelete bool) error {
	path := fmt.Sprintf("/v1/mail/messages/%s?hard_delete=%t", url.PathEscape(messageID), hardDelete)
	return m.c.do(ctx, "DELETE", path, nil, nil)
}

// ── Admin (mail:admin) ──────────────────────────────────────────────────────

func (m *MailClient) ListMailboxes(ctx context.Context, q ListMailboxesQuery) (*ListMailboxesResponse, error) {
	v := url.Values{}
	if q.Status != "" {
		v.Set("status", string(q.Status))
	}
	if q.Limit > 0 {
		v.Set("limit", strconv.Itoa(q.Limit))
	}
	if q.Cursor != "" {
		v.Set("cursor", q.Cursor)
	}
	path := "/v1/mail/mailboxes"
	if encoded := v.Encode(); encoded != "" {
		path += "?" + encoded
	}
	var out ListMailboxesResponse
	if err := m.c.do(ctx, "GET", path, nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

func (m *MailClient) CreateMailbox(ctx context.Context, req CreateMailboxRequest) (*Mailbox, error) {
	var out Mailbox
	if err := m.c.do(ctx, "POST", "/v1/mail/mailboxes", req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

func (m *MailClient) ListDomains(ctx context.Context, status DomainStatus) (*ListMailDomainsResponse, error) {
	path := "/v1/mail/domains"
	if status != "" {
		path += "?status=" + string(status)
	}
	var out ListMailDomainsResponse
	if err := m.c.do(ctx, "GET", path, nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

func (m *MailClient) CreateDomain(ctx context.Context, req CreateMailDomainRequest) (*MailDomain, error) {
	var out MailDomain
	if err := m.c.do(ctx, "POST", "/v1/mail/domains", req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// VerifyDomain re-checks DNS records for the domain. On success status flips
// to active. On failure, returns Domain with verification_errors populated.
func (m *MailClient) VerifyDomain(ctx context.Context, domainID string) (*MailDomain, error) {
	var out MailDomain
	path := fmt.Sprintf("/v1/mail/domains/%s/verify", url.PathEscape(domainID))
	if err := m.c.do(ctx, "POST", path, struct{}{}, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// RotateDKIM generates a new DKIM key for the domain. The previous selector
// remains valid for ~30 days for graceful DNS propagation. Default key_bits=2048.
func (m *MailClient) RotateDKIM(ctx context.Context, domainID string, keyBits int) (*MailDomain, error) {
	if keyBits == 0 {
		keyBits = 2048
	}
	req := RotateDKIMRequest{KeyBits: keyBits}
	var out MailDomain
	path := fmt.Sprintf("/v1/mail/domains/%s/dkim/rotate", url.PathEscape(domainID))
	if err := m.c.do(ctx, "POST", path, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}
